File: 2-hard/poker_hands/poker_hands.js
/*
* Challenge description:
* https://www.codeeval.com/open_challenges/86
*/

var fs  = require("fs");

global.ranks = {
	cardValues: {
		"2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, 
		"9": 9, "T": 10, "J": 11, "Q": 12, "K": 13, "A": 14,
	},

	combos: {
		"HighCard": 0, "OnePair": 1, "TwoPairs": 2, "ThreeOfAKind": 3, "Straight": 4, "Flush": 5, "FullHouse": 6,
		"FourOfAKind": 7, "StraightFlush": 8, "RoyalFlush": 9 
	}
}

fs.readFileSync(process.argv[2]).toString().split('\n').forEach(function (line) {
    if (line !== "") {
        cards = line.split(" ");
        console.log(compareHands(cards.slice(0, 5), cards.slice(5, 10)));
    }
});

function compareHands (leftHandCards, rightHandCards) {
	leftHand = new Hand(leftHandCards);
	rightHand = new Hand(rightHandCards);
	
	var valueLeft  = leftHand.getValue();
	var valueRight = rightHand.getValue();

	for(i = 0; i < valueLeft.length && i < valueRight.length; i++) {
		if (valueLeft[i][0] > valueRight[i][0]) {
			return 'left';
		} else if (valueLeft[i][0] < valueRight[i][0]) {
			return 'right';
		} else if (valueLeft[i][0] == valueRight[i][0]) {
			if (valueLeft[i][1] > valueRight[i][1]) {
				return 'left';
			} else if (valueLeft[i][1] < valueRight[i][1]) {
				return 'right';
			}
		}
	}

	return "none";
}

function Hand (cards) {
	this.cards = cards;
	this.outcome = [];

	this.getSortedCardValues = function () {
		var values = [];
		var cardsNum = this.cards.length;
		for (var i = 0; i < cardsNum; i++) {
			values.push(global.ranks.cardValues[this.cards[i][0]]);
		}
		return values.sort(function (a, b) { return a - b});
	}

	this.countRanks = function () {
		var cardsNum = this.cards.length;
		var rankCount = {};
		for (var i = 0; i < cardsNum; i++) {
			if (rankCount[this.cardRanks[i]]) {
				rankCount[this.cardRanks[i]] += 1;	
			} else {
				rankCount[this.cardRanks[i]] = 1;
			}
		}
		var result = {};
		for (var key in rankCount) {
			if (result[rankCount[key]]) {
				result[rankCount[key]].push(parseInt(key));
			} else {
				result[rankCount[key]] = [parseInt(key)];
			}
		}
		return result;
	}

	this.cardRanks = this.getSortedCardValues();
	this.cardSuits = this.cards.map(function (n) {return n[1];});
	this.ranksByCount = this.countRanks();
	
	this.getValue = function () {
		this.addHighCards().addPairs().addThreeOfAKind().addStraight().addFlush();
		this.addFullHouse().addFourOfAKind().addStraightAndRoyalFlush();
		return this.outcome;
	}
		
	// addRoyalFlush defined below
	// addStraightFlush defined below

	this.addFourOfAKind = function () {
		if (this.ranksByCount[4]) {
			this.outcome.unshift([global.ranks.combos.FourOfAKind, this.ranksByCount[4][0]]);
		}		

		return this;
	}

	this.addFullHouse = function () {
		if (this.ranksByCount[2] && this.ranksByCount[3]) {
			this.outcome.unshift([global.ranks.combos.FullHouse, this.ranksByCount[3][0]]);
		}
		
		return this;
	}

	this.isFlush = function () {
		var cardsNum = this.cards.length;
		for (var i = 0; i < cardsNum - 1; i++) {
			if (this.cardSuits[i] !== this.cardSuits[i + 1]) {
				return false;
			}
		}
		return true;
	}

	this.addFlush = function () {
		if (this.isFlush()) {
			this.outcome.unshift([global.ranks.combos.Flush, Math.max.apply(Math, this.cardRanks)]);	
		}
		return this;
	}

	this.isStraight = function () {
		var cardsNum = this.cards.length;
		for (var i = 0; i < cardsNum - 1; i++) {
			if (this.cardRanks[i] !== this.cardRanks[i + 1] - 1) {
				return false;
			}
		}
		return true;
	}

	this.addStraight = function () {
		if (this.isStraight()) {
			this.outcome.unshift([global.ranks.combos.Straight, Math.max.apply(Math, this.cardRanks)]);	
		}
		return this;
	}

	this.addStraightAndRoyalFlush = function () {
		if (this.isFlush() && this.isStraight()) {
			var maxRank = Math.max.apply(Math, this.cardRanks);
			this.outcome.unshift([global.ranks.combos.StraightFlush, maxRank]);		
			if (maxRank == global.ranks.cardValues['A']) {
				// royal flush
				this.outcome.unshift([global.ranks.combos.RoyalFlush, maxRank]);		
			}
		}
		return this;
	}

	this.addThreeOfAKind = function () {
		if (this.ranksByCount[3]) {
			this.outcome.unshift([global.ranks.combos.ThreeOfAKind, this.ranksByCount[3][0]]);
		}
		return this;
	}

	this.addPairs = function () {
		if (this.ranksByCount[2]) {
			//add all one pair
			for(var i = 0; i < this.ranksByCount[2].length; i++) {
				this.outcome.unshift([global.ranks.combos.OnePair, this.ranksByCount[2][i]]);
			}
			if (this.ranksByCount[2].length > 1) {
				this.outcome.unshift([global.ranks.combos.TwoPairs, Math.max.apply(Math, this.ranksByCount[2])]);
			}
		}
		
		return this;
	}

	this.addHighCards = function() {
		var comboValue = global.ranks.combos.HighCard;
		var cardsNum = this.cards.length;
		for (i = 0; i < cardsNum; i++) {
			this.outcome.unshift([comboValue, this.cardRanks[i]]);
		}
		return this;
	}
}

File: 2-hard/poker_hands/poker_hands.py
from collections import Counter
import sys

class Card:
    values = {
        '1': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7,
        '8': 8, '9': 9, 'T': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14
    }

    def __init__(self, card):
        self.rank, self.suit = card[0:2]
        self.value = Card.values[self.rank]

    def __repr__(self):
        return self.rank + self.suit

class Hand:
    def __init__(self, cards = []):
        self.cards = sorted(cards, key=lambda card: card.value, reverse=True)
        self.kind_count = self.__kind_count()

    @classmethod
    def from_string(cls, str):
        return cls(cards.Card(card) for card in str.split(' '))

    def result(self):
        return (self.highest_combo(), self.highest_rank_values())

    def highest_rank_values(self):
        return [card.value for card in self.cards]

    def highest_combo(self):
        return  (self.__royal_flush() or self.__straight_flush() or
                 self.__four_of_a_kind() or self.__full_house() or
                 self.__flush() or self.__straight() or
                 self.__three_of_a_kind() or self.__two_pairs() or
                 self.__one_pair() or self.__high_card())

    def __royal_flush(self):
        if self.__straight_flush() and self.cards[0].value == 14:
            return (9, self.cards[0].value)
        return None

    def __straight_flush(self):
        if self.__straight() and self.__flush():
            return (8, self.cards[0].value)
        return None

    def __four_of_a_kind(self):
        kind_count = self.kind_count
        if len(kind_count) == 2 and kind_count.values()[0] in (1,4):
            return (7, sorted(kind_count, key=lambda k: kind_count[k], reverse=True))
        return None

    def __full_house(self):
        kind_count = self.kind_count
        if len(kind_count) == 2 and kind_count.values()[0] in (2,3):
            return (6, sorted(kind_count, key=lambda k: kind_count[k], reverse=True))
        return None

    def __flush(self):
        for i in xrange(1, len(self.cards)):
            if self.cards[i].suit != self.cards[i - 1].suit:
                return None
        return (5, self.cards[0].value)

    def __straight(self):
        for i in xrange(1, len(self.cards)):
            if not self.cards[i].value == self.cards[i - 1].value - 1:
                return None
        return (4, self.cards[0].value)

    def __three_of_a_kind(self):
        kind_count = self.kind_count
        for card_value in kind_count:
            if kind_count[card_value] == 3:
                return (3, card_value)
        return None

    def __two_pairs(self):
        total_pairs = 0
        pair_values = []
        for i, first_card in enumerate(self.cards):
            for second_card in self.cards[i+1:]:
                if first_card.value == second_card.value:
                    total_pairs += 1
                    pair_values.append(first_card.value)
        if total_pairs == 2:
            return (2, sorted(pair_values, reverse=True))
        return None

    def __one_pair(self):
        for i, first_card in enumerate(self.cards):
            for second_card in self.cards[i+1:]:
                if first_card.value == second_card.value:
                    return (1, first_card.value)
        return None

    def __high_card(self):
        return (0, self.cards[0].value)

    def __kind_count(self):
        return dict(Counter([card.value for card in self.cards]))

with open(sys.argv[1], 'r') as f:
    for line in f:
        crds = [Card(card) for card in line.split(' ')]
        left = Hand(crds[:5]).result()
        right = Hand(crds[5:]).result()
        if left > right:
            print 'left'
        elif left < right:
            print 'right'
        else:
            print 'none'
